src/ledger/LedgerTxn.h

```diff
@@ -331,6 +331,36 @@ struct LedgerTxnDelta
     UnorderedMap<InternalLedgerKey, EntryDelta> entry;
     HeaderDelta header;
 };
+// An abstraction for storing and applying per-txn read bytes metering limits
+// when loading keys.
+class LedgerKeyMeter : public NonMovableOrCopyable
+{
+  public:
+    // Adds a transaction with a read quota and the keys it will read.
+    void addTxn(SorobanResources const& resources);
+    // Conumes entrySizeBytes from the read quotas of all transactions with this
+    // key.
+    void updateReadQuotasForKey(LedgerKey const& key, size_t entrySizeBytes);
+    // Returns true if a transaction with sufficient read quota exists for this
+    // key.
+    bool canLoad(LedgerKey const& key, size_t entrySizeBytes) const;
+    // Returns true if a key was not loaded due to insufficient read quota.
+    bool loadFailed(LedgerKey const& key) const;
+
+  private:
+    // Returns the maximum read quota across all transactions with this key.
+    uint32_t maxReadQuotaForKey(LedgerKey const& key) const;
+
+    // Stores the read quota for each transaction.
+    std::vector<uint32_t> mTxReadBytes{};
+    // Stores the keys that each transaction will read. i.e.
+    // mLedgerKeyToTxs[key] is a vector containing  the indices of
+    // mTxReadBytes corresponding to the transactions which have
+    // the key in their footprint.
+    UnorderedMap<LedgerKey, std::vector<size_t>> mLedgerKeyToTxs{};
```

Comment:
this is quite an error-prone way to store key <> tx mappings, as you have to update two data structures, and a tx is implicitly tied to the vector index (so if the index changes, it breaks the map). could we have a single map of keys mapped to {txHash, byte limit}?

Reply:
With a single map, in order to update the byte limit after reading key A I would have to iterate over all other keys, and all transactions associated with each key and update the limits separately. That's less efficient and more error prone imo

Reply:
Are envisioning something like this `map<LedgerKey, vector<pair<Hash, size_t>>` ?

Reply:
I had originally used hash, but @SirTyson suggested using indices https://github.com/stellar/stellar-core/pull/4178#discussion_r1515229783

Reply:
> This saves space, but more importantly is more straightforward, as there's some inconsistencies between using getFullHash and getContentsHash. This also let's you convert UnorderedMap<Hash, uint32_t> txReadBytes to std::vector<uint32_t> where the index into the vector corresponds to txID.

Reply:
hah :) I was thinking using hashes is lot more straightforward, since you know exactly what you're mapping. Not sure saving space matters here, we're talking a few extra bytes, right? I mean if ya'll feel strongly, lmk, just feels like there's more opportunity for bugs this way

Reply:
Its both a space saving and a (albeit small) optimization, as we have a constant index into the `vector<size_t> txReadBytes` instead of a worst-case linear lookup in the `UnorderedMap<Hash, uint32_t>`. I don't feel very strongly either way, but seeing as this is an internal data structure that is only modified indirectly by the public functions of `LedgerKeyMeter`, I don't think the current approach is that error prone. I'm happy to change it back to Hashes 🤷